lahja/trio/endpoint.py

```diff
@@ -0,0 +1,905 @@
+    Set,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+    cast,
+)
+import weakref
+
+from async_generator import asynccontextmanager
+import trio
+import trio_typing
+
+from lahja import constants
+from lahja.base import (
+    BaseEndpoint,
+    BaseRemoteEndpoint,
+    ConnectionAPI,
+    EndpointAPI,
+    RemoteEndpointAPI,
+    TResponse,
+    TStreamEvent,
+    TSubscribeEvent,
+    TWaitForEvent,
+)
+from lahja.common import (
+    BaseEvent,
+    BaseRequestResponseEvent,
+    Broadcast,
+    BroadcastConfig,
+    ConnectionConfig,
+    Message,
+    Msg,
+    Subscription,
+    should_endpoint_receive_item,
+)
+from lahja.exceptions import (
+    ConnectionAttemptRejected,
+    LifecycleError,
+    RemoteDisconnected,
+    UnexpectedResponse,
+)
+from lahja.typing import ConditionAPI, RequestID
+
+
+class TrioConnection(ConnectionAPI):
+    logger = logging.getLogger("lahja.trio.Connection")
+
+    def __init__(self, socket: trio.SocketStream) -> None:
+        self._socket = socket
+        self._msg_send_channel, self._msg_receive_channel = cast(
+            Tuple[trio.abc.SendChannel[Message], trio.abc.ReceiveChannel[Message]],
+            trio.open_memory_channel(100),
+        )
+        self._lock = trio.Lock()
+        self._running = trio.Event()
+        self._stopped = trio.Event()
+        super().__init__()
+
+    def __str__(self) -> str:
+        return f"TrioConnection[{self._socket}]"
+
+    def __repr__(self) -> str:
+        return f"<{self}>"
+
+    #
+    # Lifecycle API
+    #
+    @property
+    def is_running(self) -> bool:
+        return not self.is_stopped and self._running.is_set()
+
+    @property
+    def is_stopped(self) -> bool:
+        return self._stopped.is_set()
+
+    async def wait_started(self) -> None:
+        await self._running.wait()
+
+    async def wait_stopped(self) -> None:
+        await self._stopped.wait()
+
+    @asynccontextmanager
+    async def run(self) -> AsyncGenerator[ConnectionAPI, None]:
+        async with trio.open_nursery() as nursery:
+            await self._start(nursery)
+            try:
+                yield self
+            finally:
+                await self._stop()
+
+    async def _start(self, nursery: trio_typing.Nursery) -> None:
+        if self.is_running:
+            raise LifecycleError("Connection is already running")
+        elif self.is_stopped:
+            raise LifecycleError("Connection has already been run and stopped")
+
+        nursery.start_soon(self._run)
+        # this is a trade-off between allowing flexibility in how long it takes
+        # for a `Runnable` to start and getting good errors in the event that
+        # the `_run` method forgets to set the `_running` event.
+        with trio.fail_after(2):
+            await self.wait_started()
+
+    async def _stop(self) -> None:
+        if self.is_stopped:
+            return
+        self._stopped.set()
+        await self._cleanup()
+
+    async def _run(self) -> None:
```

Comment:
Huh, why did you decide to add this internal buffer, maybe this improved performance in the benchmark? The kernel already adds [a pretty big buffer](https://unix.stackexchange.com/questions/424380/what-values-may-linux-use-for-the-default-unix-socket-buffer-size/424381#424381) here, I was thinking about ways to simplify this code but the easiest way might be to remove it entirely and rely on the buffer that the OS provides.

Reply:
I'll try that.  I'm not entirely sure this is actually necessary so I'll verify the performance gain is real.